extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java

```diff
@@ -0,0 +1,747 @@
+   * @param name The name of the table definition to modify. The user must
+   *        have write access to the table.
+   * @param spec The new table definition.
+   * @param actionParam What to do if the table already exists.
+   *        {@code ifNew} is the same as the SQL IF NOT EXISTS clause. If {@code new},
+   *        then an error is raised if the table exists. If {@code ifNew}, then
+   *        the action silently does nothing if the table exists. Primarily for
+   *        use in scripts. The other two options are primarily for use in tests.
+   * @param req the HTTP request used for authorization.
+   * @return the version number of the table
+   */
+  @POST
+  @Path("/tables/{dbSchema}/{name}")
+  @Consumes(MediaType.APPLICATION_JSON)
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response postTable(
+      @PathParam("dbSchema") String dbSchema,
+      @PathParam("name") String name,
+      TableSpec spec,
+      @QueryParam("action") String actionParam,
+      @QueryParam("version") long version,
+      @Context final HttpServletRequest req
+  )
+  {
+    final PostAction action;
+    if (actionParam == null) {
+      action = PostAction.NEW;
+    } else {
+      action = PostAction.valueOf(StringUtils.toUpperCase(actionParam));
+      if (action == null) {
+        return Actions.badRequest(
+            Actions.INVALID,
+            StringUtils.format(
+                "Not a valid action: [%s]. Valid actions are new, ifNew, replace, force",
+                actionParam
+            )
+        );
+      }
+    }
+    TableId tableId = TableId.of(dbSchema, name);
+    Response response = authorizeTable(tableId, spec, req);
+    if (response != null) {
+      return response;
+    }
+    TableMetadata table = TableMetadata.newTable(tableId, spec);
+    try {
+      catalog.validate(table);
+    }
+    catch (IAE e) {
+      return Actions.badRequest(Actions.INVALID, e.getMessage());
+    }
+
+    switch (action) {
+      case NEW:
+        return insertTableSpec(table, false);
+      case IFNEW:
+        return insertTableSpec(table, true);
+      case REPLACE:
+        return updateTableSpec(table, version);
+      case FORCE:
+        return addOrUpdateTableSpec(table);
+      default:
+        throw new ISE("Unknown action.");
+    }
+  }
+
+  private Response authorizeTable(TableId tableId, TableSpec spec, final HttpServletRequest req)
+  {
+    // Druid has a fixed set of schemas. Ensure the one provided is valid.
+    Pair<Response, SchemaSpec> result = validateSchema(tableId.schema());
+    if (result.lhs != null) {
+      return result.lhs;
+    }
+    SchemaSpec schema = result.rhs;
+
+    // The schema has to be one that allows table definitions.
+    if (!schema.writable()) {
+      return Actions.badRequest(
+          Actions.INVALID,
+          StringUtils.format("Cannot modify schema %s", tableId.schema())
+      );
+    }
+
+    // Table name can't be blank or have spaces
```

Comment:
from logic below, it seems that the table can have spaces, just not leading or trailing spaces, is that right? If so we should update this comment.